In [2]:
#Let's do importations
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.cluster import KMeans
import numpy as np
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import mean_squared_error, accuracy_score
from sklearn.model_selection import cross_val_score,GridSearchCV,train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.feature_selection import mutual_info_classif
from sklearn.feature_selection import GenericUnivariateSelect
from sklearn.pipeline import Pipeline
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import LinearSVC, SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.naive_bayes import GaussianNB

In [3]:
#Get the dataset and display it
df = pd.read_csv("C:/Users/TUGCE/Downloads/hw5_data.csv")
display(df)

Virtual  Offset     loc       Import  Imports     var     Forwarder  \
0         3.0     3.0   689.0    18.000000      6.0   890.0      6.000000   
1         3.0     3.0   583.0    15.000000      5.0  1574.0   8640.382774   
2         3.0     4.0  2796.0    15.000000      5.0   625.0      5.000000   
3         3.0     3.0   373.0  2726.071722      9.0   615.0      9.000000   
4         3.0     3.0    77.0    12.000000      4.0    66.0  91945.635853   
...       ...     ...     ...          ...      ...     ...           ...   
4994      4.0     4.0   118.0     0.000000      3.0     0.0      0.000000   
4995      3.0     3.0    27.0    12.000000      4.0    38.0      4.000000   
4996      3.0     3.0    16.0    12.000000      4.0     7.0      4.000000   
4997      4.0     4.0  2810.0     0.000000      2.0   666.0      0.000000   
4998      3.0     4.0  1587.0    21.000000      7.0   526.0      7.000000   

      UINT          LONG          BOOL  ...          Img99         Img100  \
0     39.0     17.000000     88.000000  ... -208196.847822  149454.443336   
1      6.0      7.000000     30.000000  ...   36209.864030  221318.549792   
2     58.0     11.000000    152.000000  ...   88398.670923   66550.556919   
3     16.0  98583.277889     34.000000  ... -157101.258148  -60336.395075   
4      3.0      0.000000      5.000000  ...   32595.648968 -228909.737896   
...    ...           ...           ...  ...            ...            ...   
4994   7.0      0.000000     19.000000  ...  -73364.892264   56908.607312   
4995   3.0      0.000000  84772.903087  ...  -91323.369562  101628.992660   
4996   9.0      0.000000     16.000000  ...   26135.603443  172273.939349   
4997  14.0      9.000000     26.000000  ...  -70335.106256  -21680.148668   
4998  34.0      5.000000  61033.550970  ...  -14702.601507   18593.024103   

             Img101         Img102         Img103         Img104  \
0     330552.774213  133907.410063   44038.800343   55156.067737   
1      -1568.194718   22651.037591 -144906.975987  -33489.566102   
2       5404.362294   13947.925003  -48559.885445  257023.562444   
3     157629.928962  117458.409503   62076.273381   98733.489947   
4     -87033.363460  131606.196188 -118625.690367   89326.297602   
...             ...            ...            ...            ...   
4994   17895.279817   -4589.198675  103618.354421 -126164.290238   
4995    8812.429736  -41149.748369  -56293.578460   87524.257112   
4996   74515.593674   79555.659907   79230.809864   92047.004360   
4997   11900.591113    8015.793354 -105036.351181  -33386.286236   
4998    2249.256720   45713.524900  -16239.518388  -78553.054374   

             Img105         Img106    Img107  target  
0     -77588.974897  171979.000000  162674.0       2  
1     157701.356695    7702.000000    6551.0       8  
2    -204889.973046  151324.169975   12946.0       6  
3     -51461.636374  122247.000000   98621.0       4  
4     -84991.427204   15501.000000   11864.0       1  
...             ...            ...       ...     ...  
4994  -24272.971224    7100.000000    6002.0       4  
4995 -200433.940978   69304.000000   65638.0       3  
4996   94829.356066   71789.000000   68354.0       3  
4997  169699.674511   89989.000000   70139.0       2  
4998 -115903.637264    9394.000000    3145.0       9  

[4999 rows x 1805 columns]

In [4]:
#Create training and test set with shuffling
train_df, test_df = train_test_split(df,random_state=0) 

In [5]:
#Create an instance of the class Decision Tree Classifier, which will represent the decision tree
#I made max_depth = 3 to make test score small
tree = DecisionTreeClassifier(max_depth=3)
tree.fit(train_df.drop('target', axis=1), train_df['target'])  
print('Test set score: ', accuracy_score(test_df['target'], tree.predict(test_df.drop('target',axis=1))))

Test set score:  0.6624


To increace accuracy, using validation set instead of test set can help. One way of doing is by using cross validation. In cross validation, we will choose **k** folds, which is the number of slices we create in our dataset. We then take turns using all but one of the k folds to train on, and evaluate on the held out fold. We repeat this k times, holding out each fold once. I will apply cross validation to all classification methods that ı am evaluating.

In [6]:
#Define a method for esminating cross validation score 
#I took k as 5 for k fold cross validation in all the k fold cross validation computations
def cross_val(class_algorithm):    
    scores = cross_val_score(class_algorithm, train_df.drop('target', axis=1), train_df['target'], cv=5)
    print("Cross validation scores: ", scores)
    print("Avg Acc: " , scores.mean() , " std dev: ", scores.std(), " max: ", scores.max())
#Estimate cross validation computations for decision tree
cross_val(tree)

Cross validation scores:  [0.69333333 0.65866667 0.624      0.65066667 0.65287049]
Avg Acc:  0.6559074321317311  std dev:  0.022194973742432463  max:  0.6933333333333334


We can observe that using cross validation does help. But what about reducing the number of useless features?. This is called feature selection. Because in our dataset we have 1804 features. This is a lot. We can try different number of features and choose the one with the best cross validation score. By that way, after using feature selection we can also use cross validation and improve accuracy even more. I will apply feature selection to all classification methods that I am evaluating.

In [7]:
#Decision tree (and also random forest) already has a embedded feature selection method (feature importances)
##Define a method for calculating embedded feature selection
#Because it takes lots of time to run the code I am not iterating through every possible number of features
def feature_select_embedded (class_algorithm, new_classifier,class_name):
    ks = [i for i in range(2, df.drop('target',axis=1).shape[1],400)]
    arr = []
    for k in ks:
        features = []
        for importance, name in sorted(zip(class_algorithm.feature_importances_, df.drop('target',axis=1).columns),reverse=True)[:k]:
            features.append(name)
        x_feature = train_df[features]
        classifier_new = new_classifier
        score = cross_val_score(classifier_new, x_feature, train_df['target'], cv=5)
        print(f"For {k} Features {class_name} Mean Accuracy: ", score.mean())
        arr.append(score.mean())
    arr.sort(reverse=True)
    print("Highest score value:",arr[0])

We can observe that without feature selection mean cross validation score was 0.6553740987983978. Morever, with feature selection best mean cross validation score is 0.6556407654650644 which has feature value of 1602. It doesn't increase drastically. It is good to remember that because of running time it is not possible to iterate through every possible values of features.

In [10]:
#Call the feature selection method
new_class = DecisionTreeClassifier(max_depth=3)
dec_tree = "Decision Tree"
feature_select_embedded(tree,new_class,dec_tree)

For 2 Features Decision Tree Mean Accuracy:  0.5993612817089453
For 402 Features Decision Tree Mean Accuracy:  0.6505733867378727
For 802 Features Decision Tree Mean Accuracy:  0.6505744548286604
For 1202 Features Decision Tree Mean Accuracy:  0.6500411214953271
For 1602 Features Decision Tree Mean Accuracy:  0.6564407654650645
Highest score value: 0.6564407654650645


Now, hyperparameter tunning (finding best parameter) can increase score even more by applying grid search. I will apply hyperparameter tunning to all classification methods that I am evaluating.

In [11]:
#By finding the best feature value calculate new training and test data and fit train data into the instance of Grid Search class
features = []
for importance, name in sorted(zip(tree.feature_importances_, df.drop('target',axis=1).columns),reverse=True)[:1602]:
    features.append(name)
x_feature = train_df[features]
x_feature_test = test_df[features]
#Apply grid search to find best hyperparameter, (because it takes lots of time to run the code I am just using max depth parameter)
param_grid = [
  {
   'max_depth': [ 3, 5, 9, 15]
   },
 ]
clf = GridSearchCV(new_class, param_grid, cv=5)   
clf.fit(x_feature, train_df['target'])


GridSearchCV(cv=5, estimator=DecisionTreeClassifier(max_depth=3),
             param_grid=[{'max_depth': [3, 5, 9, 15]}])

At first, test set score was 0.6624 but now it is 0.86 by using the max depth = 9. It improved the accuracy.

In [12]:
print('Best parameter for 5 fold cross validation:',clf.best_params_)
print('Best score for a 5 fold cross validation:',clf.best_score_)
print('Test set score after CV: ', accuracy_score(test_df['target'],clf.predict(x_feature_test)))

Best parameter for 5 fold cross validation: {'max_depth': 9}
Best score for a 5 fold cross validation: 0.8375562082777035
Test set score after CV:  0.8624


In [27]:
#Create an instance of the class Decision Random Forest Classifier, which will represent the random forest classification
#But with random forest even though I try different n_estimators values the test score was always near 0.90
#This is much bigger than deicison tree's test score. TOne of the reasons it uses ensembling.It makes test score higher
ran_for=RandomForestClassifier(n_estimators=40)
ran_for.fit(train_df.drop('target', axis=1), train_df['target']) 
print('Test set score: ', accuracy_score(test_df['target'], ran_for.predict(test_df.drop('target',axis=1))))  

Test set score:  0.9064


We can observe that using cross validation does help again.    

In [28]:
cross_val(ran_for)

Cross validation scores:  [0.912      0.89733333 0.888      0.90533333 0.90520694]
Avg Acc:  0.9015747218513575  std dev:  0.008225193762418623  max:  0.912


We can observe that without feature selection mean cross validation score was 0.9015747218513575. Morever, with feature selection best mean cross validation score is 0.90424281263907454 which has feature value of 402. It doesn't increase drastically. It is good to remember that because of running time it is not possible to iterate through every possible values of features.

In [29]:
#Call the feature selection method
new_class = RandomForestClassifier(n_estimators=40)
rand_for = "Random Forest"
feature_select_embedded(ran_for,new_class,rand_for)

For 2 Features Random Forest Mean Accuracy:  0.7572713840676457
For 402 Features Random Forest Mean Accuracy:  0.9042428126390745
For 802 Features Random Forest Mean Accuracy:  0.9021087672452159
For 1202 Features Random Forest Mean Accuracy:  0.9031757899421452
For 1602 Features Random Forest Mean Accuracy:  0.9015754339118824
Highest score value: 0.9042428126390745


In [32]:
#By finding the best feature number calculate new training and test data and fit train data into the instance of Grid Search class
features = []
for importance, name in sorted(zip(ran_for.feature_importances_, df.drop('target',axis=1).columns),reverse=True)[:402]:
    features.append(name)
x_feature = train_df[features]
x_feature_test = test_df[features]

#Apply grid search to find best hyperparameter, (because it takes lots of time to run the code I am just using 1 parameter)
param_grid = [
  { 'n_estimators': [40,100,150,200], 
  },
 ]
clf = GridSearchCV(ran_for, param_grid, cv=5)   
clf.fit(x_feature, train_df['target'])  


GridSearchCV(cv=5, estimator=RandomForestClassifier(n_estimators=40),
             param_grid=[{'n_estimators': [40, 100, 150, 200]}])

At first, test set score was 0.9064 but now it is 0.908 by using the n_estimators = 150. It improved the accuracy.

In [33]:
print('Best parameter for 5 fold cross validation:',clf.best_params_)
print('Best score for a 5 fold cross validation:',clf.best_score_)
print('Test set score after CV: ', accuracy_score(test_df['target'],clf.predict(x_feature_test)))

Best parameter for 5 fold cross validation: {'n_estimators': 150}
Best score for a 5 fold cross validation: 0.9063765020026702
Test set score after CV:  0.908


In [50]:
#Create an instance of the class SVC, which will represent the SVM
#I made C = 3 to make test score small
svm = SVC(C=3)
svm.fit(train_df.drop('target', axis=1), train_df['target'])  
print('Test set score: ', accuracy_score(test_df['target'], svm.predict(test_df.drop('target',axis=1))))   # tree.score(test_df.drop('target',axis=1), test_df['target'])

Test set score:  0.4144


We can observe that using cross validation does help again.    

In [51]:
cross_val(svm)

Cross validation scores:  [0.41466667 0.404      0.41733333 0.40266667 0.4058745 ]
Avg Acc:  0.40890823319982206  std dev:  0.00593958441915745  max:  0.41733333333333333


We can observe that without feature selection mean cross validation score was 0.40890823319982206. Morever, with feature selection best mean cross validation score is 0.5993659101023587 which has feature value of 802. We can observe a good increaase.

In [52]:
##Define a method for calculating feature selection for mutual information 
#because wrapper method takes much more time than filter method I couldn't use it but it is more accurate to use wrapper method for SVM for instance
#Select the best feature by their average k fold cross validation score values
def feature_selection_mutual(class_algorithm,class_name):
  ks = [i for i in range(2, df.drop('target',axis=1).shape[1],400)]
  arr = []

  for k in ks:
    select = GenericUnivariateSelect(score_func=mutual_info_classif, mode="k_best", param=k)
    X_k = select.fit_transform(train_df.drop('target', axis=1), train_df['target'])
    score = cross_val_score(class_algorithm, X_k, train_df['target'], cv=5).mean()
    print(f"For {k} Features {class_name} Mean Accuracy: ", score)
    arr.append( score )
  arr.sort(reverse=True)
  print("Highest score value:",arr[0])

svm_name = "SVM"
feature_selection_mutual(svm,svm_name)

For 2 Features SVM Mean Accuracy:  0.43612033822874946
For 402 Features SVM Mean Accuracy:  0.5740229639519359
For 802 Features SVM Mean Accuracy:  0.5993659101023587
For 1202 Features SVM Mean Accuracy:  0.40890823319982206
For 1602 Features SVM Mean Accuracy:  0.40890823319982206
Highest score value: 0.5993659101023587


In [60]:
#By finding the best feature number calculate new training and test data and fit train data into the instance of Grid Search class
svm_new = SVC(C=3)
select = GenericUnivariateSelect(score_func=mutual_info_classif, mode="k_best", param=802)
select.fit(train_df.drop('target', axis=1), train_df['target'])
X_train = select.transform(train_df.drop('target', axis=1))
X_test = select.transform(test_df.drop('target', axis=1))

#Apply grid search to find best hyperparameter, (because it takes lots of time to run the code I am just using 1 parameter)
param_grid = [
  {
   'C': [1,3, 5, 9]
   },
 ]
clf = GridSearchCV(svm_new, param_grid, cv=5)   
clf.fit(X_train, train_df['target'])

GridSearchCV(cv=5, estimator=SVC(C=3), param_grid=[{'C': [1, 3, 5, 9]}])

At first, test set score was 0.4144 but now it is 0.6928 by using the C = 9. It improved the accuracy.

In [62]:
print('Best parameter for 5 fold cross validation:',clf.best_params_)
print('Best score for a 5 fold cross validation:',clf.best_score_)
print('Test set score after CV: ', accuracy_score(test_df['target'],clf.predict(X_test)))

Best parameter for 5 fold cross validation: {'C': 9}
Best score for a 5 fold cross validation: 0.6380414775255897
Test set score after CV:  0.6928


In [55]:
#Create an instance of the class K Neighbors Classifier, which will represent the knn
#I made n_neighbors=1, p = 2 to make test score small
knn_classifier = KNeighborsClassifier(n_neighbors=1, p=2)
knn_classifier.fit(train_df.drop('target', axis=1), train_df['target'])  
print('Test set score: ', accuracy_score(test_df['target'], knn_classifier.predict(test_df.drop('target',axis=1))))   # tree.score(test_df.drop('target',axis=1), test_df['target'])

Test set score:  0.6824


We can observe that using cross validation does help again.    

In [56]:
cross_val(knn_classifier)

Cross validation scores:  [0.68133333 0.66533333 0.67066667 0.672      0.69158879]
Avg Acc:  0.6761844236760124  std dev:  0.00926829893171791  max:  0.6915887850467289


We can observe that without feature selection mean cross validation score was 0.6761844236760124. Morever, with feature selection best mean cross validation score is 0.7311305740987983 which has feature value of 1202. We can observe a good increaase.

In [57]:
knn_name = "KNN"
feature_selection_mutual(knn_classifier,knn_name)

For 2 Features KNN Mean Accuracy:  0.6833865598575879
For 402 Features KNN Mean Accuracy:  0.7025965287049398
For 802 Features KNN Mean Accuracy:  0.650841121495327
For 1202 Features KNN Mean Accuracy:  0.7311305740987983
For 1602 Features KNN Mean Accuracy:  0.7201979528259902
Highest score value: 0.7311305740987983


In [58]:
#By finding the best feature number calculate new training and test data and fit train data into the instance of Grid Search class
knn_classifier_new = KNeighborsClassifier(n_neighbors=1, p=2)
select = GenericUnivariateSelect(score_func=mutual_info_classif, mode="k_best", param=1202)
select.fit(train_df.drop('target', axis=1), train_df['target'])
X_train = select.transform(train_df.drop('target', axis=1))
X_test = select.transform(test_df.drop('target', axis=1))

#Apply grid search to find best hyperparameter
param_grid = [
  { 'n_neighbors': [1, 2, 3, 4], 
    'p': [1, 2, 5]},
 ]
clf = GridSearchCV(knn_classifier_new, param_grid, cv=5)   
clf.fit(X_train, train_df['target'])

GridSearchCV(cv=5, estimator=KNeighborsClassifier(n_neighbors=1),
             param_grid=[{'n_neighbors': [1, 2, 3, 4], 'p': [1, 2, 5]}])

At first, test set score was 0.6824 but now it is 0.848 by using the n_neighbors = 4, p = 1. It improved the accuracy.

In [59]:
print('Best parameter for 5 fold cross validation:',clf.best_params_)
print('Best score for a 5 fold cross validation:',clf.best_score_)
print('Test set score after CV: ', accuracy_score(test_df['target'],clf.predict(X_test)))

Best parameter for 5 fold cross validation: {'n_neighbors': 4, 'p': 1}
Best score for a 5 fold cross validation: 0.8327547841566533
Test set score after CV:  0.848


In [63]:
#Create an instance of the class Guassian Naive Bayes, which will represent the naive bayes
#I made var_smoothing default to make test score small
gnb = GaussianNB()
gnb.fit(train_df.drop('target', axis=1), train_df['target'])
print('Test set score: ', accuracy_score(test_df['target'], gnb.predict(test_df.drop('target',axis=1))))   # tree.score(test_df.drop('target',axis=1), test_df['target'])

Test set score:  0.4712


We can observe that using cross validation does help again.

In [64]:
cross_val(gnb)

Cross validation scores:  [0.508      0.45866667 0.464      0.48933333 0.47396529]
Avg Acc:  0.4787930574098799  std dev:  0.017951915945319825  max:  0.508


We can observe that without feature selection mean cross validation score was 0.4787930574098799. Morever, with feature selection best mean cross validation score is 0.49586648865153543 which has feature value of 1202. We can observe an increaase.

In [65]:
gnb_name = "GNB"
feature_selection_mutual(gnb,gnb_name)

For 2 Features GNB Mean Accuracy:  0.14937676902536717
For 402 Features GNB Mean Accuracy:  0.39557703604806405
For 802 Features GNB Mean Accuracy:  0.437720694259012
For 1202 Features GNB Mean Accuracy:  0.49586648865153543
For 1602 Features GNB Mean Accuracy:  0.47666079216733426
Highest score value: 0.49586648865153543


In [46]:
#By finding the best feature number calculate new training and test data and fit train data into the instance of Grid Search class
gnb_new = GaussianNB()
select = GenericUnivariateSelect(score_func=mutual_info_classif, mode="k_best", param=1202)
select.fit(train_df.drop('target', axis=1), train_df['target'])
X_train = select.transform(train_df.drop('target', axis=1))
X_test = select.transform(test_df.drop('target', axis=1))

#Apply grid search to find best hyperparameter, (because it takes lots of time to run the code I am just using 1 parameter)
param_grid= {'var_smoothing': np.logspace(0,-9, num=100)}
clf = GridSearchCV(gnb_new, param_grid, cv=5)   
clf.fit(X_train, train_df['target'])

GridSearchCV(cv=5, estimator=GaussianNB(),
             param_grid={'var_smoothing': array([1.00000000e+00, 8.11130831e-01, 6.57933225e-01, 5.33669923e-01,
       4.32876128e-01, 3.51119173e-01, 2.84803587e-01, 2.31012970e-01,
       1.87381742e-01, 1.51991108e-01, 1.23284674e-01, 1.00000000e-01,
       8.11130831e-02, 6.57933225e-02, 5.33669923e-02, 4.32876128e-02,
       3.51119173e-02, 2.84803587e-02, 2.31...
       1.23284674e-07, 1.00000000e-07, 8.11130831e-08, 6.57933225e-08,
       5.33669923e-08, 4.32876128e-08, 3.51119173e-08, 2.84803587e-08,
       2.31012970e-08, 1.87381742e-08, 1.51991108e-08, 1.23284674e-08,
       1.00000000e-08, 8.11130831e-09, 6.57933225e-09, 5.33669923e-09,
       4.32876128e-09, 3.51119173e-09, 2.84803587e-09, 2.31012970e-09,
       1.87381742e-09, 1.51991108e-09, 1.23284674e-09, 1.00000000e-09])})

At first, test set score was 0.4712 but now it is 0.5856 by using the var_smoothing = 6.579332246575682e-07. It improved the accuracy.

In [47]:
print('Best parameter for 5 fold cross validation:',clf.best_params_)
print('Best score for a 5 fold cross validation:',clf.best_score_)
print('Test set score after CV: ', accuracy_score(test_df['target'],clf.predict(X_test)))

Best parameter for 5 fold cross validation: {'var_smoothing': 6.579332246575682e-07}
Best score for a 5 fold cross validation: 0.5673513128615932
Test set score after CV:  0.5856


I will use majority voting to see if it would be better to create new classifier from extisting classifiers. I will use the classification methods that has the same feature values. And first fit them with their best parameter values that I have found.

In [66]:
#For majority voting I am going to use two classification methods (K Nearest Neighborhood and Naive Bayes)  
#From the feature selection above we can observe that they show the best performance for the same value of features
#I will also use the above Grid Search best parameter results and we will se if score result will increase or not

from sklearn.ensemble import VotingClassifier
classifiers = [GaussianNB(var_smoothing=6.579332246575682e-07), KNeighborsClassifier(n_neighbors=4, p=1)]
names = ["Naive Bayes" ,"4-NN"]
classifiers.append( VotingClassifier([ (type(x).__name__, x) for x in classifiers ]) )
names.append("Vote(NB,5-NN)")

Formula of voting classifier is: $$h_\text{ensemble}(x) = \frac{1}{z} \sum_{i=1}^z h_i(x)$$ Because 4NN score is much more than NB score when we take voting classifier ensemble of this 2 classification methods the score result becomes less than the score result of 4NN. Thus, in this case it is better to choose 4NN.

In [67]:
for model, m_name in zip(classifiers, names):
  model.fit(X_train, train_df['target'])
  print("\t", m_name, accuracy_score(test_df['target'], model.predict(X_test)))

	 Naive Bayes 0.452
	 4-NN 0.8152
	 Vote(NB,5-NN) 0.6712


As a result is better to use Random Forest classification method because it is accuracy is about 90% since the beginning.